Define PaymentRequestEvent.respondWith() behavior.

romandev opened the pull request:
The current spec says outline of the method's behavior but it is not specific.
This patch makes the behavior of the method clearer. (#190, #191)

romandev committed:
Define PaymentRequestEvent.respondWith() behavior.

The current spec says outline of the method's behavior but it is not specific.
This patch makes the behavior of the method clearer. (#190, #191)

romandev commented:
@ianbjacobs, @rsolomakhin PTAL

ianbjacobs commented:
Hi @romandev,

This looks like very good work. I feel I am not the best person to review this and so would like to hear from @rsolomakhin and anyone else who has more experience on these paradigms. Thank you!

Ian

rsolomakhin commented on index.html:
```diff
@@ -1918,40 +1957,110 @@
       </section>
       <section>
         <h2>
-          Extention to User Accepts the Payment Request Algorithm
```
Should the spec say somewhere that "Respond to PaymentRequest Algorithm" should be running as part of the "User Accepts the Payment Request Algorithm" in https://w3c.github.io/payment-handler/?

rsolomakhin replied:
Found it below.

rsolomakhin reviewed (approved):
LGTM

ianbjacobs commented:
@mkruisselbrink, as this relates to the issue you raised (#190) it would be great to hear whether you feel this change addresses your issue (and thus, if merged, can we close #190?).

Thank you,

Ian

mkruisselbrink commented on index.html:
```diff
@@ -1918,40 +1957,110 @@
+          <li>If <var>event</var>'s <a>dispatch flag</a> is unset, then
+          <a>throw</a> an "<a>InvalidStateError</a>" <a>DOMException</a> and
+          abort these steps.
           </li>
```
Should there be a check here, or somewhere else, to make sure respondWith is only called once (i.e. similar to FetchEvent's "respond-with entered flag")?

mkruisselbrink commented on index.html:
```diff
@@ -1918,40 +1957,110 @@
-          [[!payment-request]], replacing steps 6 and 7 with these steps or
-          their equivalent:
+          When this algorithm with <var>event</var> and
+          <var>handlerResponsePromise</var> parameters are invoked, the user
```
This should at least be "When this algorithm [...] is invoked", and maybe even re-order it to "When this algorithm is invoked with `event` and `handlerResponsePromise` parameters, "

mkruisselbrink commented on index.html:
```diff
@@ -1918,40 +1957,110 @@
+                      "PaymentResponse"><var>response</var></a>.<a data-cite=
+                      "!payment-request#dom-paymentresponse-methodname">methodName</a>.
+                      </li>
+                      <li>Create a <a>structured clone</a> of
```
"structured clone" does not exist anymore in the latest HTML spec (because the way it was defined was problematic with which realms things are done in). Instead you'll need to call [StructuredSerialize](http://w3c.github.io/html/infrastructure.html#section-structuredserialize) in the context of the service worker (i.e. before calling the "user accepts the payment request algorithm"), and then [StructuredDeserialize](http://w3c.github.io/html/infrastructure.html#section-structureddeserialize) here, to create javascript objects in the correct realm. One reason for that is because serialization can have observable side effects in the source context, so it needs to be well specified when serialization happens.

mkruisselbrink commented on index.html:
```diff
@@ -1918,40 +1957,110 @@
-          <li>Create a <a>structured clone</a> of
-          <var>handlerResponse</var>.<var>details</var> and assign it to <var>
-            response</var>.<var>details</var>.
+          <li>Perform the remaining steps in parallel:
```
I don't think you need "perform the remaining steps in parallel", or the "wait until promis settles" below. Instead you can just directly have 
1. Upon rejection of handlerResponsePromise:
   1. Do stuff
1. Upon fulfillment of handlerResponsePromise:
   1. Do stuff
1. Upon fulfillment or rejection of handlerResponsePromise queue a microtask to perform the following steps:
   1. Do stuf

mkruisselbrink reviewed (commented):
This definitely looks a lot better than what was there before. There still is the weirdness of saying that the user agent MUST run the "payment app failure" algorithm, without any requirement as to what that algorithm does. It's not really a MUST requirement if there aren't any actual requirements.

marcoscaceres reviewed (dismissed):
Needs tests

romandev commented:
I'll update this PR in this week. (including tests if possible)

romandev reviewed (commented):
Rebased and updated.

ianbjacobs commented:
@mkruisselbrink  and @romandev are you satisfied with the state of this pull request? Thanks!

Ian

romandev commented:
@ianbjacobs Sorry for delayed reply.
I'm attending BlinkOn8 in Tokyo without note pc.
Anyway I addressed all comments except payment app failure algorithm.
The algorithm is unclear now but we can improve it in follow-up PR.
Thanks.
